### PR TITLE
Add a --no-security flag

### PR DESCRIPTION
**Pull request by mia-pi-git**
Not sure if i like how the start script handles this.

### CODE REVIEW

**Comment by Slayer95** [comment]
Every decision that decreases security should be a conscious choice and absolutely cannot be implemented through generic catch-all flags. If someone needs to regularly dismiss several security features they should write their own bash script to exec PS with their required args.

File: pokemon-showdown [hunk 1]
```diff
@@ -86,8 +86,8 @@ if (!process.argv[2] || /^[0-9]+$/.test(process.argv[2])) {
 	case 'start':
 		{
 			process.argv.splice(2, 1);
-			if (process.argv[2] === '--skip-build') {
-				process.argv.splice(2, 1);
+			if (process.argv.includes('--skip-build')) {
+				process.argv.splice(process.argv.indexOf('--skip-build'), 1);
```
Comment (Zarel) [inline pokemon-showdown:90]: You don't have the splice for `--no-security`. I suspect it might not be necessary.

Comment (Zarel) [reply pokemon-showdown:90]: Probably should have all the argv's in one place?

**Comment by Zarel** [comment]
> Every decision that decreases security should be a conscious choice and absolutely cannot be implemented through generic catch-all flags. If someone needs to regularly dismiss several security features they should write their own bash script to exec PS with their required args.

To be clear, none of the options disabled here are particularly dangerous. They're all ones you want disabled while doing local development testing. The name is intentionally a bit scary so people don't try turning it on their prod servers, but they're things I think a wide variety of people would want disabled for test use-cases.

File: server/config-loader.ts [hunk 2]
```diff
@@ -33,6 +37,12 @@ export function load(invalidate = false) {
 		}
 	}
 
+	for (const [preset, values] of FLAG_PRESETS) {
+		if (process.argv.includes(preset)) {
+			for (const value of values) config[value] = true;
+		}
+	}
```
Comment (Zarel) [inline server/config-loader.ts:44]: How is the port detected? That's probably relevant if you're not going to remove the settings from `process.argv`.

Comment (mia-pi-git) [reply server/config-loader.ts:44]: I'd rather just remove the flag from process.argv and let port be handled normally, if I'm understanding you correctly?

Comment (mia-pi-git) [reply server/config-loader.ts:44]: a9b4303 should do it.

Comment (Zarel) [reply server/config-loader.ts:44]: I don't like how you do the removal and detection in separate loops. It's easy enough just to skip everything starting with `-` when detecting the port.

Comment (mia-pi-git) [reply server/config-loader.ts:44]: There, does that work better?

File: server/index.ts [hunk 3]
```diff
@@ -163,7 +163,11 @@ if (require.main === module) {
 	// in the case of app.js being imported as a module (e.g. unit tests),
 	// postpone launching until app.listen() is called.
 	let port;
-	if (process.argv[2]) port = parseInt(process.argv[2]);
+	for (const arg of process.argv) {
+		const parsed = parseInt(arg);
+		if (isNaN(parsed)) continue;
+		port = parsed;
```
Comment (Zarel) [inline server/index.ts:169]:
```suggestion
		if (/^[0-9]+$/.test(arg)) {
			port = parseInt(arg);
			break;
		}
```

Remember, `parseInt('98.3 oifajew')` is `98`.

**Comment by Zarel** [comment]
@Slayer95 I _am_ open to a different name, though. Something like "dev-mode".

@xfix @monsanto do either of you think this is a bad idea for security reasons?

**Comment by Slayer95** [comment]
Of course the name "no-security" doesn't make any sense. "Dev-Mode", on the other hand, doesn't convey anything about what it actually does, and it makes it seem as if this flag had no risks other than maybe stability, or performance degradation due to profiling (which the developer might expect if they turn on a "dev mode").

I shall also take the chance to point out that the setting ``noguestsecurity`` can easily become an attack vector against bots that may connect to such a server and accept console commands from impersonators. And yes, I know for certain that there are plenty of bots out there which would be vulnerable.

**Comment by Slayer95** [comment]
Now, if we go back to the drawing board, and consider the use case:

> they're all ones you want disabled while doing local development testing

Then shouldn't we look for a setting which deals with that more literally such as some "trust-loopback", which disables checks for connections from ``127.0.0.1/8`` or ``::1/128`` ?

**Comment by Zarel** [comment]
> Then shouldn't we look for a setting which deals with that more literally such as some "trust-loopback", which disables checks for connections from `127.0.0.1/8` or `::1/128` and also sets the bind address?

My concern is that you might very well want to test e.g. a bot hosted on a different server. I don't see a need for finer-grained permissions when they're only used in development test servers and none of them are particularly dangerous in that context; I think it's more worthwhile to have one setting that immediately flips all this stuff for anything that doesn't test security.

**Comment by xfix** [comment]
I don't think we should trust loopback without requiring a server owner to explicilty pick such an option, you may as well host a Showdown server with, say, nginx proxy pass (I know I did that back when I hosted my server).

I like the name `--no-security` myself, makes it rather clear you shouldn't use it in production.

**Comment by Slayer95** [comment]
> you may as well host a Showdown server with, say, nginx proxy pass

PS supports ``X-Forwarded-For`` through ``Config.noproxy`` though, and sets the connection ``ip`` accordingly.

As for the rest of both your comments, guess we can only disagree. ``</thread>``

**Comment by xfix** [comment]
> PS supports `X-Forwarded-For` through `Config.noproxy` though, and sets the connection `ip` accordingly.

That's assuming you know about it, I know I personally ran a server for a month before realizing this option exists as misconfiguring it isn't immediately obvious, and even if you do notice everybody having `::1` as an IP address doesn't mean it's clear what needs to be changed. Also note that `config-example.js` discourages changing the value of this option.

```javascript
/**
 * proxyip - proxy IPs with trusted X-Forwarded-For headers
 *   This can be either false (meaning not to trust any proxies) or an array
 *   of strings. Each string should be either an IP address or a subnet given
 *   in CIDR notation. You should usually leave this as `false` unless you
 *   know what you are doing
 * @type {false | string[]}.
 */
exports.proxyip = false;
```

That being said, personally I'm not a fan of `noguestsecurity`. If you want to test stuff with unregistered users, you may as well type `/nick ` and smash random letters on keyboard (this is very likely to be an unregistered account), and if you need to test something where permissions matter, well, this won't help with that.

The other options are fine with me, and definitely something Showdown should support.

**Comment by Zarel** [comment]
> That being said, personally I'm not a fan of `noguestsecurity`. If you want to test stuff with unregistered users, you may as well type `/nick ` and smash random letters on keyboard (this is very likely to be an unregistered account), and if you need to test something where permissions matter, well, this won't help with that.

This is mostly for `poke-env`, which wants to write AI battle bots without going through the hassle of a login handshake.

We could maybe require the username prefix "Guest", for slightly more security?

**Comment by xfix** [comment]
Fair then.

**Comment by Zarel** [comment]
I imagine `noguestsecurity` would also be useful to others who want to write unit tests for their bots or clients, but don't want the tests to require network use (it's also possible to generate your own login keypair, but that's undocumented and a lot more effort).

**Comment by Zarel** [comment]
So we decided against #7690. I do like the untrusted symbol. Maybe a permanent `!` trumping all other symbols, in front of untrusted users?

Untrusted users being marked however, maybe `user.trusted === false`?

**Comment by hsahovic** [comment]
What is the status on this PR?
I tested the config values that were introduced beforehand (`nothrottle`, `noguestsecurity` and `noipchecks`) and they are performing as well as expected :)

**Comment by Zarel** [comment]
I have some nitpicks but it's worth merging as-is and nitpicks can be fixed later.

**Comment by Zarel** [comment]
Thanks, Mia!

**Pull request merged**